rockcraft/services/lifecycle.py

```diff
@@ -0,0 +1,98 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2023 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Rockcraft Lifecycle service."""
+
+from __future__ import annotations
+
+import contextlib
+from pathlib import Path
+from typing import cast
+
+from craft_application import LifecycleService
+from craft_archives import repo
+from craft_cli import emit
+from craft_parts import callbacks
+from craft_parts.errors import CallbackRegistrationError
+from overrides import override
+
+from rockcraft.models.project import Project
+
+
+class RockcraftLifecycleService(LifecycleService):
+    """Rockcraft-specific lifecycle service."""
+
+    @override
+    def setup(self) -> None:
+        """Configure extra args to the LifecycleManager."""
+        # pylint: disable=import-outside-toplevel
+        # This inner import is necessary to resolve a cyclic import
+        from rockcraft.services import RockcraftServiceFactory
```

Comment:
Would it work if we import at toplevel, but protected with `TYPE_CHECKING`?

Reply:
Interesting. If I do that, `pylint` complains that it doesn't know what `RockcraftServiceFactory` in the `cast()` call. However, if I do `cast("RockcraftServiceFactory", self._services)` it seems to work??

Reply:
Maybe this solves the cyclic import warning too.

Reply:
Just checked, it doesn´t.